content/post/gld.md

```diff
@@ -0,0 +1,1355 @@
+  | cut -d/ -f2 | sort | uniq | wc -l
+$ # boyter/lc
+$ time lc . \
+  | grep -vE 'NOASSERTION|----|Directory' | cut -d" " -f1 | sort | uniq | wc -l
+```
+
+## Algorithm
+
+We have implemented license detection based on the license and README files for now, and wish to
+add fine-grained scanning of source code files in the future (do you want to help us?). Given the
+stated license text, we compare it to the texts in the SPDX database and record the match. The naive
+way of making the comparisons is to calculate the
+[Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)
+between the query and each of the examples. This is not fast at all for the two reasons:
+
+1. The complexity is linear to the number of examples.
+2. The complexity is linear to the size of the difference. Since we've always got very few matches,
+the rest of the distances will be slow to calculate.
+
+These two reasons confidently render the naive approach unusable.
+
+The core of go-license-detector's detection mechanism is [Locality Sensitive Hashing](https://en.wikipedia.org/wiki/Locality-sensitive_hashing).
+We treat each license as a TF-IDF-weighted bag-of-words - that is, as a set of words where every
+word has a weight corresponding to the frequency in the license text (term frequency) and
+throughout the whole database (document frequency). This is a proven approach for large scale
+similarity detection which we used multiple times in the past. Although 400 items is clearly
+not large scale at all, it still makes sense to employ LSH because of the O(1) complexity
+guarantee. We saw that it works reasonably well in practice and introduces the minimal overhead,
+say 20MB of memory for the hashes and the vocabulary. The hashing algorithm which we use is
+[Weighted MinHash](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/36928.pdf)
+- again, battle-tested in the past, e.g. in [Apollo](https://github.com/src-d/apollo) or
+[bags deduplication](https://blog.sourced.tech/post/minhashcuda/).
+We set 75% Jaccard similarity threshold and hash length 154
+after careful tuning of the false positive vs. false negative vs. performance trade-off.
+Since we discard the text structure by treating sequences as sets, we further calculate the Levenshtein
+distance to the matched database records in order to determine the precise confidence value.
+
+We look at the README file if the analyzed project does not contain a license file. This happens
+in more than 7% of the cases in the 1k dataset and x% in Public Git Archive (182,000 repositories).
+There is a fair chance that the license name is mentioned in the README, so we apply
+[Named Entity Recognition](https://en.wikipedia.org/wiki/Named-entity_recognition)
+to find them using the excellent [jdkato/prose](https://github.com/jdkato/prose) NLP library for Go.
+
+Unfortunately, the devil is not in the core algorithm but rather in numerous details which cover
```

Comment:
I like the phrase, but use it on its original form for more impact.

> Unfortunately, the devil is in the details. There are many unexpected variations for which we had to modify
> our initially straight-forward algorithm.

Or something like that?